### PR TITLE
#240 Added gutter route navigation from/to Java/XML

### PR DESCRIPTION
**Pull request by adrianbumbas**
First version of the navigation that suports Java literals and XML only. The issue #242 is still not solved so i did a workaround to filter out everything that is not a literal for Java.

### CODE REVIEW

**Comment by adrianbumbas** [comment]
@davsclaus feel free to merge this if you don't have any comments

File: camel-idea-plugin/src/main/java/org/apache/camel/idea/gutter/CamelRouteLineMarkerProvider.java [hunk 1]
```diff
@@ -69,4 +93,116 @@ private CamelPreferenceService getCamelPreferenceService() {
...
+     *
+     * @param route      the complete Camel route to search for
+     * @param psiElement the {@link PsiLiteralExpression} that might contain the complete route definition
+     * @return the {@link PsiElement} that contains the exact match of the Camel route, null if there is no exact match
+     */
+    private PsiElement findJavaElement(String route, PsiLiteralExpression psiElement) {
+        if (route.equals(psiElement.getValue())) {
+            //the method 'to' is a PsiIdentifier not a PsiMethodCallExpression because it's part of method invocation chain
+            PsiMethodCallExpression methodCall = PsiTreeUtil.getParentOfType(psiElement, PsiMethodCallExpression.class);
+            if (methodCall != null && "to".equals(methodCall.getMethodExpression().getReferenceName())) {
```
Comment (davsclaus) [inline camel-idea-plugin/src/main/java/org/apache/camel/idea/gutter/CamelRouteLineMarkerProvider.java]: You can also have `toD` or `toF` in Camel. The former is used more than the latter which is a shortcut for using String.format. to build the uri string.

File: camel-idea-plugin/src/main/java/org/apache/camel/idea/gutter/CamelRouteLineMarkerProvider.java [hunk 2]
```diff
@@ -69,4 +93,116 @@ private CamelPreferenceService getCamelPreferenceService() {
...
+    /**
+     * Further refine search in order to match the exact XML Camel route.
+     *
+     * @param route      the complete Camel route to search for
+     * @param psiElement the {@link PsiElement} that might contain the complete route definition
+     * @return the {@link PsiElement} that contains the exact match of the Camel route
+     */
+    private PsiElement findXMLElement(String route, XmlToken psiElement) {
+        if (psiElement.getTokenType() == XmlElementType.XML_ATTRIBUTE_VALUE_TOKEN) {
+            if ("to".equals(PsiTreeUtil.getParentOfType(psiElement, XmlTag.class).getLocalName())) {
```
Comment (davsclaus) [inline camel-idea-plugin/src/main/java/org/apache/camel/idea/gutter/CamelRouteLineMarkerProvider.java]: In XML DSL there is only `<to>` and <toD>`

File: camel-idea-plugin/src/main/java/org/apache/camel/idea/gutter/CamelRouteLineMarkerProvider.java [hunk 3]
```diff
@@ -69,4 +93,116 @@ private CamelPreferenceService getCamelPreferenceService() {
...
+     * Defines the search scope for the Camel routes.
+     * It consist of all the files from all the modules including the files from JAR files.
+     *
+     * @param project the current {@link Project}
+     * @return the search scope
+     */
+    private GlobalSearchScope getSearchScope(Project project) {
+        Module[] modules = ModuleManager.getInstance(project).getModules();
+
+        GlobalSearchScope searchScope = GlobalSearchScope.EMPTY_SCOPE;
```
Comment (davsclaus) [inline camel-idea-plugin/src/main/java/org/apache/camel/idea/gutter/CamelRouteLineMarkerProvider.java]: Are we not searching to far? I think you should only search in the project source code, eg `src/main/java` and maybe also `src/test/java` etc. Not in JARs and other libraries.

File: camel-idea-plugin/src/test/resources/testData/JavaCamelRouteLineMarkerProviderTestData.java [hunk 4]
```diff
@@ -1,38 +1,29 @@
...
 
-public final class JavaCamelRouteLineMarkerProviderTestData {
-
-    private JavaAddEndpointIntentionTestData() { }
-
-    public static void main(String[] args) {
-        Main main = new Main();
-        main.addRouteBuilder(new MyRouteBuilder());
-        main.run(args);
+public class JavaCamelRouteLineMarkerProviderTestData extends RouteBuilder {
```
Comment (davsclaus) [inline camel-idea-plugin/src/test/resources/testData/JavaCamelRouteLineMarkerProviderTestData.java]: I think this PR has checkstyle errors, when you build the code it should report errors.

**Comment by davsclaus** [comment]
@adrianbumbas lets merge this and then you can do another PR with the minor improvements I pointed out. 

It would be good to get a release out with latest stuff, and then refine and get community feedback

**Comment by adrianbumbas** [comment]
@davsclaus thanks, i will do the requested changes.

**Pull request merged**